### PR TITLE
MULE-8192: Ensure that WSConsumer disposes the outbound endpoints it cre...

### PR DESCRIPTION
**Pull request by marcosnc**
...ates. Ensure that outbound endpoints removes themselves from the connector's dispatchers cache.

### CODE REVIEW

File: core/src/main/java/org/mule/endpoint/AbstractEndpoint.java [hunk 1]
```diff
@@ -508,6 +508,12 @@ public void dispose()
         // this.messageProcessors.clear();
         // Don't clear this, since it changes the hash code, which can foul up shutdown processing
         // when objects have been keyed by endpoint, e.g. dispatchers
+
```
Comment (pablokraan) [inline core/src/main/java/org/mule/endpoint/AbstractEndpoint.java]: Put the new lines before the comment, so the comment is right next to the lines where it applies

File: modules/ws/src/test/java/org/mule/module/ws/consumer/WSConsumerTestCase.java [hunk 2]
```diff
@@ -66,6 +76,22 @@ public void failsToInitializeWithInvalidOperation() throws MuleException
...
+    public void outboundEndpointIsDisposed() throws Exception
+    {
+        TestEndpointFactory endpointFactory = new TestEndpointFactory();
+        muleContext.getRegistry().registerObject(OBJECT_MULE_ENDPOINT_FACTORY, endpointFactory);
+
+        WSConsumer wsConsumer = createConsumer();
+        wsConsumer.getConfig().setConnector(new HttpConnector(muleContext));
+        wsConsumer.initialise();
+
+        verify((Disposable) endpointFactory.getCreatedEndpoint(), never()).dispose();
```
Comment (pablokraan) [inline modules/ws/src/test/java/org/mule/module/ws/consumer/WSConsumerTestCase.java]: this line is redundant at this point

File: modules/ws/src/test/java/org/mule/module/ws/consumer/WSConsumerTestCase.java [hunk 3]
```diff
@@ -84,4 +110,23 @@ private WSConsumer createConsumer()
 
         return wsConsumer;
     }
+
+    class TestEndpointFactory extends DefaultEndpointFactory
```
Comment (pablokraan) [inline modules/ws/src/test/java/org/mule/module/ws/consumer/WSConsumerTestCase.java]: make it private

File: modules/ws/src/test/java/org/mule/module/ws/consumer/WSConsumerTestCase.java [hunk 4]
```diff
@@ -84,4 +110,23 @@ private WSConsumer createConsumer()
...
+    {
+
+        private DefaultOutboundEndpoint createdEndpoint;
+
+        @Override
+        public OutboundEndpoint getOutboundEndpoint(EndpointBuilder builder) throws MuleException
+        {
+            createdEndpoint = mock(DefaultOutboundEndpoint.class);
```
Comment (pablokraan) [inline modules/ws/src/test/java/org/mule/module/ws/consumer/WSConsumerTestCase.java]: Move the field initialization to the field definition

File: core/src/main/java/org/mule/endpoint/AbstractEndpoint.java (hunk 1 above)
Comment (marcosnc) [inline core/src/main/java/org/mule/endpoint/AbstractEndpoint.java]: The comments above are related to the commented clear() on the messageProcessors field, they are previous to this change and not related to it. We could move and separate the comments in order to make them clearer.

File: modules/ws/src/test/java/org/mule/module/ws/consumer/WSConsumerTestCase.java (hunk 2 above)
Comment (marcosnc) [inline modules/ws/src/test/java/org/mule/module/ws/consumer/WSConsumerTestCase.java]: You're right, I only wanted to be sure that dispose were not called as a side effect of another operation.

File: modules/ws/src/test/java/org/mule/module/ws/consumer/WSConsumerTestCase.java (hunk 3 above)
Comment (marcosnc) [inline modules/ws/src/test/java/org/mule/module/ws/consumer/WSConsumerTestCase.java]: Ok.

File: modules/ws/src/test/java/org/mule/module/ws/consumer/WSConsumerTestCase.java (hunk 4 above)
Comment (marcosnc) [inline modules/ws/src/test/java/org/mule/module/ws/consumer/WSConsumerTestCase.java]: Ok.

**Pull request closed without merging**